ycmd/completers/completer.py

```diff
@@ -556,6 +560,60 @@ def GetSettings( self, module, request_data ):
     return {}
 
 
+  def Reset( self ):
+    self._server_file_state = completer_utils.ServerFileStateStore()
+
+
+  def OpenFileHandler( self, file_state, filetypes, changes ):
+    pass # pragma: no cover
+
+
+  def ChangeFileHandler( self, file_state, filetypes, changes ):
+    pass # pragma: no cover
+
+
+  def _AnySupportedFileType( self, file_types ):
+    for supported in self.SupportedFiletypes():
+      if supported in file_types:
+        return True
+    return False
+
+
+  def _UpdateDirtyFilesUnderLock( self, request_data ):
+    for file_name, file_data in request_data[ 'file_data' ].items():
+      if not self._AnySupportedFileType( file_data[ 'filetypes' ] ):
+        LOGGER.debug( 'Not updating file %s, it is not a supported filetype: '
+                       '%s not in %s',
+                       file_name,
+                       file_data[ 'filetypes' ],
+                       self.SupportedFiletypes() )
+        continue
+
+      file_state = self._server_file_state[ file_name ]
+      if 'contents' in file_data:
+        action = file_state.GetDirtyFileAction( file_data[ 'contents' ] )
+      elif 'changes' in file_data:
+        changes = file_data[ 'changes' ]
+        changes_start = changes[ 'start' ][ 'line_num' ]
+        changes_end = changes[ 'end' ][ 'line_num' ]
+        file_state[ changes_start :
+                    changes_end ] = changes[ 'replacement_text' ]
+        action = completer_utils.ServerFileState.CHANGE_FILE
```

Comment:
how do we know this is the right action ? I mean if the client sent changes without contents, we're assuming we already sent a didOpen ? Is that safe?

Reply:
Also, when we do this, what's updating the version number in `ServerFileState` ?